core/tx_pool.go

```diff
@@ -211,6 +221,165 @@ func (config *TxPoolConfig) sanitize() TxPoolConfig {
+
+// Close wraps corresponding method on the websocket but is safe for concurrent calling
+func (wsConnection *wsConn) CloseWS() error {
+	// The Close and WriteControl methods can be called concurrently with other methods
+	return wsConnection.conn.Close()
+}
+
+// Go routine that looks for messages from the relay
+func (pool *TxPool) readWSMessages() {
+	for {
+		messageType, message, err := pool.wsConnection.ReadMessage()
+		if err != nil {
+			if messageType == -1 { // message type emitted when relay ws connection is closed
+				log.Error("WS connection with relay closed")
+				pool.wsConnection.wsConnected = false
+			} else {
+				log.Error("WS error while reading the relay message: ", err.Error(), messageType)
+			}
+			return
+		}
+		var abstractMessage relayAbstractMessage
+		if err := json.Unmarshal([]byte(message), &abstractMessage); err != nil {
+			log.Error("Error while decoding relay message: ", err.Error(), nil)
+			return
+		}
+		// If relay message is of type "success", log the success message from relay
+		if abstractMessage.Type == "success" {
+			var successMessage relaySuccessMessage
+			if err := json.Unmarshal([]byte(message), &successMessage); err != nil {
+				log.Error("Error while decoding relay success message: ", err.Error(), nil)
+				return
+			}
+			log.Info(successMessage.Data)
+		}
+		// If relay message is of type "bundle", decode the payload and add bundle
+		// The sender is responsible for signing the transaction and using the correct nonce and ensuring validity
+		if abstractMessage.Type == "bundle" {
+			var bundleMessage relayBundleMessage
+			if err := json.Unmarshal([]byte(message), &bundleMessage); err != nil {
+				log.Error("Error while decoding relay bundle message: ", err.Error(), nil)
+				return
+			}
+
+			var txs types.Transactions
+
+			for _, encodedTx := range bundleMessage.Data.EncodedTxs {
+				tx := new(types.Transaction)
+				if err := rlp.DecodeBytes(encodedTx, tx); err != nil {
+					return
+				}
+				txs = append(txs, tx)
+			}
+			// Finally, we add the bundle sent to the tx pool
+			pool.AddMevBundle(txs, big.NewInt(bundleMessage.Data.BlockNumber.Int64()), bundleMessage.Data.MinTimestamp, bundleMessage.Data.MaxTimestamp, bundleMessage.Data.RevertingTxHashes)
+		}
+
+	}
+}
+
+// Go routine to ping the ws server periodically to ensure a live connection
+func (pool *TxPool) ping() {
+	pool.wsConnection.wlock.Lock()
+	defer pool.wsConnection.wlock.Unlock()
+	ticker := time.NewTicker(pool.wsConnection.pingPeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if !pool.wsConnection.wsConnected && pool.wsEnabled {
+				// If a ws connection hasn't already been established, attempt it
+				pool.connectWS()
+			} else {
+				// If it has already been connected, test with a ping message
+				if err := pool.wsConnection.conn.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(pool.wsConnection.pingPeriod)); err != nil {
+					pool.wsConnection.wsConnected = false
```

Comment:
Does it take up to `wsConnection.pingPeriod` time for a re-connection to occur? Would it be better to try reconnecting immediately?

Reply:
Good catch, it waits 10 seconds but I changed this to attempt reconnection immediately 
```
if err := pool.wsConnection.conn.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(pool.wsConnection.pingPeriod)); err != nil {
    pool.wsConnection.wsConnected = false
    log.Info("Error while sending a ping to the relay WS, attempting reconnection")
    pool.connectWS()
```

Reply:
The ` time.Now().Add(pool.wsConnection.pingPeriod)` in this context is the deadline for the ping message (client expects a response within 10 seconds or raises an error)